### PR TITLE
Feature/proper naming and legacy

### PR DESCRIPTION
**Pull request by ruseinov**
This includes full refactoring of all legacy wraps. After this is merged all the broken tests need to be fixed and that'd be it.

Note that we start errors with code 1, because otherwise tendermint `ResponseCheckTx` that check CodeTypeOk (which is 0) will fail to pass and error will be treated as ok. 

0 is reserved for non-errors.

### CODE REVIEW

File: errors/errors.go [hunk 1]
```diff
@@ -2,42 +2,41 @@ package errors
...
-
 	"github.com/pkg/errors"
 )
 
 var (
 	// InternalErr represents a general case issue that cannot be
 	// categorized as any of the below cases.
-	InternalErr = RegisterNew(0, "internal")
+	// We start as 1 as 0 is reserved for non-errors
+	InternalErr = Register(1, "internal")
```
Comment (husio) [inline errors/errors.go]: Why did you change the internal error to 1?

The design of errors was relying on the internal error code being a zero value. This is important for the non weave errors, because they do not explicitly provide an error code.

File: x/batch/msg.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ type Msg interface {
 func Validate(msg Msg) error {
 	l, err := msg.MsgList()
 	if err != nil {
-		return errors.Wrap(err)
+		return errors.Wrap(err, "error retrieving batch messages")
```
Comment (husio) [inline x/batch/msg.go]: I think "cannot retrieve batch message" might be a better phrasing. :nail_care:

Comment (ethanfrey) [reply x/batch/msg.go]: I agree with these naming suggestions, also above.
If we return an error, the text "error" is pretty unnecessary....
"error file not found" or "file not found"?

File: errors/errors.go (hunk 1 above)
Comment (ethanfrey) [reply errors/errors.go]: This is a good question.
My only hard requirements here is that the abci interface (deliver/checktx) *must* return code 0 on success and code > 0 on failure, as Roman referred to in the comment.

That said, I will make a suggestion below

Comment (ethanfrey) [reply errors/errors.go]: The error is returned here, as well as identical code for CheckTx:


```go
func DeliverTxError(err error, debug bool) abci.ResponseDeliverTx {
	tm := errors.Wrap(err)
	log := tm.ABCILog()
	if debug {
		log = fmt.Sprintf("%v", tm)
	}

	return abci.ResponseDeliverTx{
		Code: tm.ABCICode(),
		Log:  log,
	}
}
```

And I think Piotr's three points above are the three ways to override.
I think (1) which would change the `Code: tm.ABCICode(),` line is a bit more fragile, as we much make sure to update in all (two) places... 

I would fix it in the Wrap function. Also, note that this has no description and will delegate to the `deprecatedLegacyWrap`, which does already use 1 as a default for internal errors.

Updating `wrappedError.ABCICode` is as easy as updating the value for internal errors as Roman did. I think this is fine in all cases as is.

```go
func (e *wrappedError) ABCICode() uint32 {
	if e.Parent == nil {
		return InternalErr.code
	}
	type coder interface {
		ABCICode() uint32
	}
	if p, ok := e.Parent.(coder); ok {
		return p.ABCICode()
	}
	return InternalErr.code
}
```

So.... in conclusion, I think Roman's approach of changing the error code is proper, but we clearly need:
*  a test case of handling both wrapped and stdlib errors in `DeliverTxError` (ensure code always proper)
* reconsider why we use `tm := errors.Wrap(err)` and thus `deprecatedLegacyWrap` in `DeliverTxError`

File: errors/errors.go [hunk 3]
```diff
@@ -78,23 +77,12 @@ func (e Error) New(description string) error {
...
-	}
-	return wrapng(err, strings.Join(description, ", "))
-}
 
 // Wrap extends given error with an additional information.
 //
 // If the wrapped error does not provide ABCICode method (ie. stdlib errors),
 // it will be labeled as internal error.
-func wrapng(err error, description string) TMError {
+func Wrap(err error, description string) TMError {
```
Comment (ethanfrey) [inline errors/errors.go]: Note that the "legacy" Wrap also handles nil error... which is quite useful to avoid if statements everywhere. If you add

```go
if err == nil {
  return nil
}
```

to the top of this function, this makes it more flexible. Note the conversation here: https://github.com/iov-one/weave/pull/305#discussion_r255043779

Something like

```golang
if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
  return fmt.Errorf("cannot broadcast transaction: %s", err)
}
return nil
```

in old code, could be simplified to

```golang
return errors.Wrap(
  bnsClient.BroadcastTx(&tx).IsError(), 
  "cannot broadcast transaction"))
```

Whether you want to use such a shortcut is a personal decision based on where it is being returned, but I think it is quite helpful for the library to allow such usage

File: x/batch/msg.go (hunk 2 above)
Comment (ruseinov) [reply x/batch/msg.go]: yeah, same, this was just a placeholder to make the whole refactoring work

**Comment by ruseinov** [comment]
LGTM
